A43-sdk-authorization.md

```diff
@@ -0,0 +1,721 @@
+      string.
+   3. If we have a client certificate, we check against certificate contents. A
+      match is found if principal matches URI SANs from the certificate. If
+      there are no URI SANs in certificate, or no match was found, we check
+      against DNS SANs. Similarly, if certificate has no DNS SANs or match wasn't
+      found, we check against Subject field from certificate.
+   
+   Consider the case where the *principals* list is empty, then we only check for
+   step 1 above, i.e. rule matches for any authenticated user.
+3. Each *request* could contain a list of URL *paths* (i.e. fully qualified RPC
+   methods) and list of http *headers* to match. Refer JSON schema above to
+   understand matching semantics.
+   - If the sub-fields are empty, like if the *paths* is empty (assuming *headers*
+     is unset), the behavior would be similar to an empty *request* since no other
+     *request* fields are set.
+
+#### Example
+
+In the following policy example
+- Peer identity from ["spiffe://foo.com/sa/admin1", "spiffe://foo.com/sa/admin2"] is 
+  authorized to access any RPC methods in pkg.service
+- Any authenticated user is allowed to access "foo" and "bar" RPC methods if the
+  HTTP header includes a name "dev-path" with prefix value "dev/path/".
+- Nobody can access "secret" RPC method, not even the admins as deny rules are evaluated 
+  first.
+
+```json
+{
+	"name": "example-policy",
+	"allow_rules": [
+		{
+			"name": "admin-access",
+			"source": {
+				"principals": [
+					"spiffe://foo.com/sa/admin1",
+					"spiffe://foo.com/sa/admin2"
+				]
+			},
+			"request": {
+				"paths": ["/pkg.service/*"]
+			}
+		},
+		{
+			"name": "dev-access",
+			"source": {
+				"principals": []
+			},
+			"request": {
+				"paths": [
+					"/pkg.service/foo",
+					"/pkg.service/bar"
+				],
+				"headers": [
+					{
+						"key": "dev-path",
+						"values": ["/dev/path/*"]
+					}
+				]
+			}
+		}
+	],
+	"deny_rules": [
+		{
+			"name": "deny-access",
+			"request": {
+				"paths": [
+					"*/secret"
+				]
+			}
+		}
+	]
+}
+```
+
+### API
+
+gRPC will support both static initialization and dynamically reloading the policy
+from filesystem. In static initialization, the policy will be provided as a JSON
+string. In dynamic file reloading, the application will specify the file path that
+contains the authorization policy in JSON format.
+
+Valid user provided SDK authorization policy creates authorization engine(s).
+In the case of file watcher, we internally create thread(C-core)/ goroutine(Go)/
+scheduled service(Java) which will be used to read the policy file periodically,
+and update the authorization engines. During the first file read, if the policy
+is invalid or there are I/O errrors, we will return error back to application and
+SDK authorization won't be enabled. If the error occurs on a later reload, then
+that particular reload will be skipped, and we will continue to use the latest
```

Comment:
nit. Should we also notify the application somehow that the reload fails (maybe via log)?

Reply:
Yes, we do have logging on failed reload. Added to the proposal.